### ISSUE TITLE
in operator on `T & object` narrows to `never`

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#50639]
# Bug Report

### 🔎 Search Terms

- in operator
- generic
- intersection
- object
- never

### 🕗 Version & Regression Information

- This changed between versions 4.7 and 4.8

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/MYewdgzgLgBAZjAvDAPAQQDQD4AUA3AQwBsBXAUwC4Y0BKJLGAbwCgY2YBLBHV9vqAJ4AHMiASFSZJImQAiEACMAVmWBRZMAGSbefNhPIwAhDJhgSRIrph0We9gbIBua3wD0bmAD0A-NYC+zP4uzKCQsHAATEiomLiOVGhaMIoqanSIDHbsXDA89myyQgBOIEIaHGAwjta2rg7E5C4FMB7efnyBwcyh4NDwAMwx6Nj4jZTUGVnWufkFgiJi1ePScqmq6lo6LY7GpuaWW-WFJWUVVTV8dTvjzQVtvsetngCiAB4iamQAJjAABklNCllBs-jAFCRYABzECwP5gMh4MjFP4BIJOIA)

### 💻 Code

```ts
const f3 = <A,>(value: A) => {
    if (
        typeof value === "object" &&
        value !== null &&
        "prop" in value
    ) {
        value;
        // ^?
        // TS 4.7: `A`
        // TS 4.8: `never`
        // Expected: `A & object`
    }
};
```

### 🙁 Actual behavior

See above.

### 🙂 Expected behavior

See above.

---

**Comment by erheron** [comment]
It's quite interesting though, since if you remove the `"prop" in value` condition `if`, `value` is inferred as `A & object` as expected 🤔

**Comment by OliverJAsh** [comment]
Note that even in TS 4.7 using the `in` operator on `A & object` returned `never`: https://www.typescriptlang.org/play?ts=4.7.4#code/MYewdgzgLgBAZgJhgXhgHgIIBoB8AKANwEMAbAVwFMAuGDGAMhhACMArC4KAShRxgG8AUDBEwAlnBh5ho2QCIADgCcQCueLAxi5CjJE8hs2dsoBuPUYD0lmAD0A-BYC+gp6aA. This only became an issue after TS 4.8 started narrowing `A` to `A & object`.

**Comment by fatcerberus** [comment]
Since this involves narrowing via `in`, this might be related to https://github.com/microsoft/TypeScript/issues/50531.

**Comment by andrewbranch** [comment]
Yes, related to #50531, but not going to be fixed for 4.8.3. 4.8.3 will contain a more minimal fix specific to `unknown` and `{}`, while for 4.9 we plan to fix the other examples you show here. @ahejlsberg is actively working on this. I’m not sure there’s an existing issue tracking it though, so I guess this is it now.
